aten/src/ATen/native/cuda/Copy.cu

```diff
@@ -283,6 +319,50 @@ static bool copy_requires_temporaries(TensorIterator& iter, bool p2p_enabled) {
   }
 
   bool same_dtype = iter.dtype(0) == iter.dtype(1);
+#if defined(__CUDACC__) && !defined(__HIP__)
+  bool is_complex = at::isComplexType(iter.dtype(1));
+  bool broadcast =  broadcast_required(iter.tensor(0), iter.tensor(1));
+
+  if (iter.ndim() == 2 && same_dtype && !broadcast && !is_complex) {
+    const auto& src_tensor = iter.tensor(1);
+    const auto& dst_tensor = iter.tensor(0);
+
+    size_t element_size = src_tensor.element_size();
+    int64_t dim0 = src_tensor.size(0);
+    int64_t dim1 = iter.ndim() == 1 ? 1 : src_tensor.size(1);
```

Comment:
this is wrong, iter.ndim() and tensor dimensions don't necessarily correspond 1-to-1, because iterator constructor does dimension permutation and coalescing to improve memory accesses. In particular, in the failing test the tensors are 
```
sizes0  [3, 3, 3]
strides0 [12, 1, 4]
sizes1 [3, 3, 3]
strides1 [9, 1, 3]
```